Headscale/Tailscale tutorial (#3615)
galaxyproject/training-material: topics/admin/tutorials/wireguard-headscale/tutorial.md

---
layout: tutorial_hands_on

title: "Deploying Tailscale/Headscale for private mesh networking"
zenodo_link: ""
questions:
  - What is Tailscale?
  - When is it useful?
  - Is it right for me?
objectives:
  - Setup a tailnet across a few nodes
time_estimation: "60m"
key_points:
  - Tailscale is a fantastic bit of software that Just Works™
  - We use headscale, an open source reimplementation of Tailscale's control server because it's easy to use in training
  - But if you can afford Tailscale, just use that.
  - There is a FOSS plan, go check it out!
contributions:
  authorship:
  - hexylena
requirements:
  - type: "internal"
    topic_name: admin
    tutorials:
      - ansible
  - type: "none"
    title: "Three or more VMs (they can be tiny, 1 CPU, <1GB RAM)"
subtopic: cloud
tags:
  - wireguard
  - networking
---

[Tailscale](https://tailscale.com/) makes secure networking easy, it really is like magic. If you've used wireguard before, you know it takes a bit to setup and some configuration if you need to do anything fancy.

> ### Agenda
>
> 1. TOC
> {:toc}
>
{: .agenda}

# What is Tailscale?

> It's like wireguard but easier and they built a **lot** of nice features on top of it. It's networking that "Just Works™" even more than Wireguard.

Reviewer: ```suggestion
It's like Wireguard but easier and they built a **lot** of nice features on top of it. It's networking that "Just Works™" even more than Wireguard. If you prefer to use plain Wireguard without Headscale/Tailscale, or just want to get an understanding of the technology that Headscale/Tailscale build off of, [there is a tutorial for that]({% link topics/admin/tutorials/wireguard/tutorial.md %}) as well.
```

>    >
>    > > ### {% icon code-out %} Output: Bash
>    > >
>    > > Your hostname is probably different:
>    > > <div class="Headscale" markdown="1">
>    > > Pick one to be the `head`, and the rest to be the `tail`. The head will act as the coordination server, the `tail` will be the nodes that should talk to each other.
>    > > {% raw %}
>    > > ```diff
>    > > --- /dev/null
>    > > +++ b/hosts
>    > > @@ -0,0 +1,5 @@
>    > > +[head]
>    > > +1-wg.galaxy.training
>    > > +[tail]
>    > > +2-wg.galaxy.training
>    > > +3-wg.galaxy.training
>    > > +4-wg.galaxy.training
>    > > {% endraw %}
>    > > ```
>    > > {: data-commit="Add hosts"}
>    > > </div>
>    > >
>    > > <div class="Tailscale" markdown="1">
>    > > Place all of the nodes in a `tail` group
>    > > ```diff
>    > > --- /dev/null
>    > > +++ b/hosts
>    > > @@ -0,0 +1,5 @@
>    > > +[tail]
>    > > +1-wg.galaxy.training
>    > > +2-wg.galaxy.training
>    > > +3-wg.galaxy.training
>    > > +4-wg.galaxy.training
>    > > ```
>    > > </div>
>    > {: .code-out}
>    {: .code-2col}
{: .hands_on}

<div class="Headscale" markdown="1">

> ### {% icon details %} Consider using Tailscale

> > Tailscale has implemented the coordination server and infrastructure with much better, more robust infrastructure that you won't have to be responsible for. If you can get your university to pay for it, or grant money covering it, it's **probably worth it**. They add many new features often, and things like the mobile apps only work with their service.

Reviewer: ```suggestion
> Tailscale has implemented the coordination server and infrastructure with much better, more robust infrastructure that you won't have to be responsible for. If you can get your institution to pay for it, or grant money covering it, it's **probably worth it**. They add many new features often, and things like the mobile apps only work with their service. Tailscale is free for personal use (i.e. to test things out) and offers a free plan for open source projects that you may qualify for.
```

>

> > For a training event obviously we want something free and quick to setup and destroy, so, we're using [Headscale](https://github.com/juanfont/headscale) since it's free and we're just going to destroy it immediately, and no one will accidentally get billed ;)

Reviewer: ```suggestion
> For a training event obviously we want something free and quick to setup and destroy, so, we're using [Headscale](https://github.com/juanfont/headscale) since it's free and we're just going to destroy it immediately, and no one will accidentally get billed ;)
>
> Using Headscale will also teach you everything you need to know if you do choose to use Tailscale, which is simpler and has fewer components for you to manage yourself.
```
